### ISSUE TITLE
Support for PHP 8

### ISSUE THREAD

**Issue by cleptric** [swig/swig#1929]
Are there any plans for SWIG supporting PHP 8 yet?

**Comment by yarda** [comment]
We have PHP 8 in Fedora Rawhide (f35) which means we need to disable PHP bindings which are built with SWIG.

**Comment by ojwb** [comment]
If you try what SWIG generates for PHP7 what happens?

I don't see anything about changes to PHP's C API in https://www.php.net/manual/en/migration80.php

**Comment by geographika** [comment]
From comments on the MapServer project at https://github.com/MapServer/MapServer/issues/6201#issuecomment-748019666 - looks like there could be an issue around `ZEND_ARG_PASS_INFO(0)`
Possibly related? https://bugs.php.net/bug.php?id=80469

**Comment by jplesnik** [comment]
> 
> And there are some more errors like e.g.:
> 
> ```
> error: 'ZEND_ARG_PASS_INFO' was not declared in this scope; did you mean 'ZEND_ARG_OBJ_INFO'
> ```

The error is probably related to this changes https://github.com/php/php-src/commit/ac4e0f0852ce780e143013ceff45067a172e8a83#diff-bb1285146d0d0b77580bccc637afc7da6446eb0d3ad06925cc3a056c53de8a98L100

**Comment by ojwb** [comment]
I really wish PHP would have decent docs for their C API, especially around what changed between major releases.

I'm very short of SWIG hacking time right now, but have had a quick poke.

The version check is easy to relax.

We can generate arginfo blocks with parameter names (it sounds like `ZEND_ARG_PASS_INFO` was removed accidentally, but I don't see a way to specify `NULL` for the parameter name without it, short of assuming the `struct` layout).  That should also work for PHP7.

And it looks like we now need `HashTable * ht = Z_OBJ_HT_P(z)->get_properties(Z_OBJ_P(z));` - not checked if that also works with PHP7 but we can conditionalise on the version if necessary.

With those fixed, some new issues are uncovered:

* Warnings like this for every testcase (which presumably would be fixed by just specifying suitable arginfo for these helpers):
```
Warning: Missing arginfo for swig_unions_alter_newobject() in Unknown on line 0
Warning: Missing arginfo for swig_unions_get_newobject() in Unknown on line 0
```

* Testcase `li_math` gives `Warning: Constant M_E already defined in /home/olly/git/swig/Examples/test-suite/php/li_math.php on line 22` and similar - I think this is a new warning in PHP 8 as the testcase isn't new and PHP defined most of these constants at least as far back as PHP 5 (and probably further) but we didn't get these warnings before PHP 8.  I'm rather dubious of the utility of `Lib/math.i` since I'd expect any general purpose language to already provide access to such maths functions and constants anyway (and likely under the same names as C so these wrappers are liable to cause issues by colliding with or replacing the built-in versions.)

*
```
checking php testcase director_overload2
Fatal error: Declaration of OverloadDerived1::nnn($vvv) must be compatible with OverloadBase::nnn($vvv = null) in /home/olly/git/swig/Examples/test-suite/php/director_overload2.php on line 92
```

There may be further issues not yet visible, but the SWIG testsuite is pretty thorough and most of the testcases are now passing (warnings aside.)

If anyone else is wants to help with this I can push a branch with a tidied up version of what I have so far. Otherwise I'll chip away when I have the time.

**Comment by neoclust** [comment]
Hi,

i am following it as in Mageia 8 is using php 8 by default.

I think that a branch allowing to build with php8 ( w/o the configure check ) could help possible contributors to help.

**Comment by ojwb** [comment]
Tidying it up to push would use some of my very limited SWIG-hacking time - while possible contributors remain hypothetical it's more efficient if I just chip away at it.  But I think I'm getting close to having something generally useful.

> I really wish PHP would have decent docs for their C API, especially around what changed between major releases.

My wish is at least partly granted!  I've since found this:

https://github.com/php/php-src/blob/PHP-8.0.4/UPGRADING.INTERNALS

It doesn't seem to be linked to from anywhere obvious - I stumbled upon it while searching in the PHP source code for something else.
